apps/backend/src/modules/openai/openai.service.ts

```diff
@@ -0,0 +1,141 @@
+import OpenAI from "openai";
+
+import { config } from "~/libs/modules/config/config.js";
+
+import {
+	MAX_ATTEMPTS,
+	ONE,
+	SECONDS_TO_MILLISECONDS,
+	TEMPERATURE,
+	ZERO,
+} from "./libs/constants/constants.js";
+import { OpenAIRoles, PlanErrorMessages } from "./libs/enums/enums.js";
+import { type GeneratedPlanDTO } from "./libs/types/types.js";
+import {
+	generateAssistantPrompt,
+	systemPrompt,
+} from "./libs/utilities/utilities.js";
+import {
+	planCreateValidationSchema,
+	taskCreateValidationSchema,
+} from "./libs/validation-schemas/validation-schemas.js";
+
+class OpenAIService {
+	private maxAttempts: number;
+	private openai: OpenAI;
+
+	public constructor() {
+		this.openai = new OpenAI({
+			apiKey: config.ENV.OPEN_AI.OPEN_AI_KEY,
+		});
+		this.maxAttempts = MAX_ATTEMPTS;
+	}
+
+	public async generatePlan(
+		userPrompt: string,
+		attempts: number = ZERO,
+		assistantPrompt?: string,
+	): Promise<GeneratedPlanDTO> {
+		let answer = "";
+
+		try {
+			answer = await this.callAPI({
+				assistantPrompt,
+				systemPrompt,
+				userPrompt,
+			});
+
+			const planData = JSON.parse(answer) as GeneratedPlanDTO;
+
+			this.validatePlan(planData);
+
+			return planData;
+		} catch (error) {
+			if (attempts + ONE < this.maxAttempts) {
+				await new Promise((resolve) =>
+					setTimeout(resolve, SECONDS_TO_MILLISECONDS * attempts),
+				);
+
+				const errorMessage =
+					error instanceof Error ? error.message : String(error);
+
+				return await this.generatePlan(
+					userPrompt,
+					attempts + ONE,
+					generateAssistantPrompt(errorMessage, answer),
+				);
+			}
+
+			throw new Error(PlanErrorMessages.GENERATION_FAILED);
+		}
```

Comment:
```suggestion
			if (attempts + ONE >= this.maxAttempts) {
             throw new Error(PlanErrorMessages.GENERATION_FAILED);
			}
     // Here is the code that you put inside if clause previously
```
Let's change this condition block into early quit.